crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2384 @@
+ * @param ctx is the pointer for the HPKE context
+ * @param ikme is a buffer for the IKM
+ * @param ikmelen is the length of the above
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_ikme(OSSL_HPKE_CTX *ctx,
+                            const unsigned char *ikme, size_t ikmelen)
+{
+    if (ctx == NULL || ikme == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    OPENSSL_free(ctx->ikme);
+    ctx->ikme = OPENSSL_malloc(ikmelen);
+    if (ctx->ikme == NULL)
+        return 0;
+    memcpy(ctx->ikme, ikme, ikmelen);
+    ctx->ikmelen = ikmelen;
+    return 1;
+}
+
+/**
+ * @brief set a private key for HPKE authenticated modes
+ * @param ctx is the pointer for the HPKE context
+ * @param privp is an EVP_PKEY form of the private key
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_authpriv(OSSL_HPKE_CTX *ctx, EVP_PKEY *privp)
+{
+    if (ctx == NULL || privp == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->mode != OSSL_HPKE_MODE_AUTH
+        && ctx->mode != OSSL_HPKE_MODE_PSKAUTH) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->authpriv != NULL)
+        EVP_PKEY_free(ctx->authpriv);
+    ctx->authpriv = EVP_PKEY_dup(privp);
+    if (ctx->authpriv == NULL)
+        return 0;
+    return 1;
+}
+
+/**
+ * @brief set a public key for HPKE authenticated modes
+ * @param ctx is the pointer for the HPKE context
+ * @param pub is an buffer form of the public key
+ * @param publen is the length of the above
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_authpub(OSSL_HPKE_CTX *ctx,
+                               const unsigned char *pub, size_t publen)
+{
+    if (ctx == NULL)
+        return 0;
+    if (ctx->authpub != NULL)
+        OPENSSL_free(ctx->authpub);
+    ctx->authpub = OPENSSL_malloc(publen);
+    if (ctx->authpub == NULL)
+        return 0;
+    memcpy(ctx->authpub, pub, publen);
+    ctx->authpublen = publen;
+    return 1;
+}
+
+/**
+ * @brief ask for the state of the sequence of seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq returns the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The value returned is the most recent used when sealing
+ * or opening (successfully)
+ */
+int OSSL_HPKE_CTX_get0_seq(OSSL_HPKE_CTX *ctx, uint64_t *seq)
+{
+    if (ctx == NULL || seq == NULL)
+        return 0;
+    *seq = ctx->seq;
+    return 1;
+}
+
+/**
+ * @brief set the sequence value for seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq set the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The value returned is the most recent used when sealing
+ * or opening (successfully)
+ */
+int OSSL_HPKE_CTX_set1_seq(OSSL_HPKE_CTX *ctx, uint64_t seq)
+{
+    if (ctx == NULL)
+        return 0;
+    ctx->seq = seq;
```

Comment:
Maybe this should error if you try to go backwards (i.e potentially reuse the seq).. It also need a pod file staement saying that this is mainly for testing.

Reply:
Not sure. If some application was likely to see lost packets and wanted to recover then this could be used for more than just testing. So I think I'd argue to leave as-is.

Reply:
add a note to POD that this could leave to nonce-reuse for GCM which is v. v. bad

Reply:
done